src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMap.java

```diff
@@ -13,212 +13,644 @@
 package io.reactivex.internal.operators.flowable;
 
 import java.util.Queue;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.*;
 
 import org.reactivestreams.*;
 
-import io.reactivex.Flowable;
 import io.reactivex.functions.Function;
-import io.reactivex.internal.queue.*;
+import io.reactivex.internal.functions.Objects;
+import io.reactivex.internal.fuseable.QueueSubscription;
+import io.reactivex.internal.queue.SpscArrayQueue;
 import io.reactivex.internal.subscriptions.*;
-import io.reactivex.internal.util.Pow2;
+import io.reactivex.internal.util.Exceptions;
 import io.reactivex.plugins.RxJavaPlugins;
-import io.reactivex.subscribers.SerializedSubscriber;
-
-public final class FlowableConcatMap<T, U> extends Flowable<U> {
-    final Publisher<T> source;
-    final Function<? super T, ? extends Publisher<? extends U>> mapper;
-    final int bufferSize;
-    public FlowableConcatMap(Publisher<T> source, Function<? super T, ? extends Publisher<? extends U>> mapper, int bufferSize) {
-        this.source = source;
-        this.mapper = mapper;
-        this.bufferSize = bufferSize;
+
+public final class FlowableConcatMap<T, R> extends FlowableSource<T, R> {
+
+    final Function<? super T, ? extends Publisher<? extends R>> mapper;
+    
+    final int prefetch;
+    
+    final ErrorMode errorMode;
+    
+    /**
+     * Indicates when an error from the main source should be reported.
+     */
+    public enum ErrorMode {
```

Comment:
Enum is not a problem here, these are not part of a binary flag unlike the fusion constants.